File: src/commands/filings.ts
import TurndownService from 'turndown';
import { gfm } from '@joplin/turndown-plugin-gfm';

import { CLIError, ErrorCode } from '../core/errors.js';
import { CommandContext, CommandResult } from '../core/runtime.js';
import { filingDocumentUrl, submissionsUrl } from '../sec/endpoints.js';
import { dateInRange, normalizeAccession } from '../sec/normalizers.js';
import { resolveEntity } from '../sec/ticker-map.js';

interface RecentFilings {
  accessionNumber?: string[];
  filingDate?: string[];
  reportDate?: string[];
  acceptanceDateTime?: string[];
  act?: string[];
  form?: string[];
  fileNumber?: string[];
  filmNumber?: string[];
  items?: string[];
  size?: number[];
  isXBRL?: number[];
  isInlineXBRL?: number[];
  primaryDocument?: string[];
  primaryDocDescription?: string[];
}

interface SubmissionsPayload {
  cik: string;
  name?: string;
  tickers?: string[];
  filings?: {
    recent?: RecentFilings;
  };
}

interface FilingRow {
  accession: string;
  form: string | null;
  filingDate: string | null;
  reportDate: string | null;
  primaryDocument: string | null;
  filingUrl: string | null;
}

function buildMarkdownConverter(): TurndownService {
  const service = new TurndownService({
    headingStyle: 'atx',
    hr: '---',
    bulletListMarker: '-',
    codeBlockStyle: 'fenced',
    fence: '```',
    emDelimiter: '*',
    strongDelimiter: '**',
    linkStyle: 'inlined'
  });

  service.use(gfm);
  service.remove(['script', 'style', 'noscript', 'iframe', 'canvas']);

  return service;
}

const markdownConverter = buildMarkdownConverter();

function stripInlineXbrlHeaders(content: string): string {
  return content
    .replace(/<ix:header[\s\S]*?<\/ix:header>/gi, '')
    .replace(/<ix:hidden[\s\S]*?<\/ix:hidden>/gi, '')
    .replace(/<ix:resources[\s\S]*?<\/ix:resources>/gi, '');
}

function splitMarkdownTableCells(line: string): string[] {
  const trimmed = line.trim();
  const withoutLeadingPipe = trimmed.startsWith('|') ? trimmed.slice(1) : trimmed;
  const withoutTrailingPipe = withoutLeadingPipe.endsWith('|')
    ? withoutLeadingPipe.slice(0, -1)
    : withoutLeadingPipe;

  return withoutTrailingPipe.split('|').map((cell) => cell.trim());
}

function isMarkdownTableSeparatorLine(line: string): boolean {
  const cells = splitMarkdownTableCells(line);
  if (cells.length === 0) {
    return false;
  }

  return cells.every((cell) => /^:?-{3,}:?$/.test(cell.replace(/\s+/g, '')));
}

function collapseLayoutTables(markdown: string): string {
  const lines = markdown.split('\n');
  const output: string[] = [];

  for (let idx = 0; idx < lines.length; idx += 1) {
    const line = lines[idx];
    if (!line.trimStart().startsWith('|')) {
      output.push(line);
      continue;
    }

    const tableBlock = [line];
    while (idx + 1 < lines.length && lines[idx + 1].trimStart().startsWith('|')) {
      idx += 1;
      tableBlock.push(lines[idx]);
    }

    const hasSeparator = tableBlock.some(isMarkdownTableSeparatorLine);
    if (!hasSeparator) {
      output.push(...tableBlock);
      continue;
    }

    const dataRows = tableBlock.filter((row) => !isMarkdownTableSeparatorLine(row));
    const nonEmptyCellCounts = dataRows.map(
      (row) => splitMarkdownTableCells(row).filter((cell) => cell.length > 0).length
    );
    const maxNonEmptyCells = Math.max(...nonEmptyCellCounts, 0);
    const avgNonEmptyCells =
      nonEmptyCellCounts.reduce((sum, count) => sum + count, 0) /
      Math.max(nonEmptyCellCounts.length, 1);
    const isLayoutTable = maxNonEmptyCells <= 1 || avgNonEmptyCells <= 1.2;

    if (!isLayoutTable) {
      output.push(...tableBlock);
      continue;
    }

    const flattenedRows = dataRows
      .map((row) => splitMarkdownTableCells(row).filter((cell) => cell.length > 0).join(' '))
      .map((row) => row.replace(/\s+/g, ' ').trim())
      .filter((row) => row.length > 0);

    if (flattenedRows.length > 0) {
      output.push(...flattenedRows, '');
    }
  }

  return output.join('\n').replace(/\n{3,}/g, '\n\n').trim();
}

function zipRecentFilings(cik: string, recent: RecentFilings | undefined): FilingRow[] {
  if (!recent) {
    return [];
  }

  const accessionNumbers = recent.accessionNumber ?? [];
  const forms = recent.form ?? [];
  const filingDates = recent.filingDate ?? [];
  const reportDates = recent.reportDate ?? [];
  const primaryDocuments = recent.primaryDocument ?? [];

  const rowCount = accessionNumbers.length;
  const rows: FilingRow[] = [];

  for (let idx = 0; idx < rowCount; idx += 1) {
    const accession = accessionNumbers[idx];
    if (!accession) {
      continue;
    }

    const primaryDocument = primaryDocuments[idx] ?? null;
    const filingUrl =
      primaryDocument && primaryDocument.length > 0
        ? filingDocumentUrl({
            cik,
            accession,
            primaryDocument
          })
        : null;

    rows.push({
      accession,
      form: forms[idx] ?? null,
      filingDate: filingDates[idx] ?? null,
      reportDate: reportDates[idx] ?? null,
      primaryDocument,
      filingUrl
    });
  }

  return rows;
}

function extractMarkdownFromHtml(content: string): string {
  const sanitizedHtml = stripInlineXbrlHeaders(content);
  const markdown = markdownConverter
    .turndown(sanitizedHtml)
    .replace(/\u00a0/g, ' ')
    .replace(/\r/g, '')
    .replace(/[ \t]+\n/g, '\n')
    .replace(/\n[ \t]+/g, '\n')
    .replace(/\n{3,}/g, '\n\n')
    .trim();

  return collapseLayoutTables(markdown);
}

export async function runFilingsList(
  params: {
    id: string;
    form?: string;
    from?: string;
    to?: string;
    queryLimit?: number;
    offset?: number;
  },
  context: CommandContext
): Promise<CommandResult> {
  const entity = await resolveEntity(params.id, context.secClient, { strictMapMatch: false });

  const submissions = await context.secClient.fetchSecJson<SubmissionsPayload>(submissionsUrl(entity.cik));
  const rows = zipRecentFilings(entity.cik, submissions.filings?.recent);

  const normalizedForm = params.form?.toUpperCase();
  const filteredRows = rows.filter((row) => {
    if (normalizedForm && (row.form ?? '').toUpperCase() !== normalizedForm) {
      return false;
    }

    if (!row.filingDate) {
      return !params.from && !params.to;
    }

    return dateInRange(row.filingDate, params.from, params.to);
  });

  const offset = params.offset ?? 0;
  const queryLimit = params.queryLimit ?? filteredRows.length;
  const pagedRows = filteredRows.slice(offset, offset + queryLimit);

  return {
    data: pagedRows,
    metaUpdates: {
      query_total_count: filteredRows.length,
      query_returned_count: pagedRows.length,
      query_truncated: offset + pagedRows.length < filteredRows.length,
      query_offset: offset
    }
  };
}

export async function runFilingsGet(
  params: {
    id: string;
    accession: string;
    format: 'url' | 'html' | 'text' | 'markdown';
  },
  context: CommandContext
): Promise<CommandResult> {
  const accession = normalizeAccession(params.accession);
  const entity = await resolveEntity(params.id, context.secClient, { strictMapMatch: false });

  const submissions = await context.secClient.fetchSecJson<SubmissionsPayload>(submissionsUrl(entity.cik));
  const rows = zipRecentFilings(entity.cik, submissions.filings?.recent);
  const match = rows.find((row) => row.accession === accession);

  if (!match) {
    throw new CLIError(
      ErrorCode.NOT_FOUND,
      `Accession ${accession} not found in recent submissions for ${params.id}`
    );
  }

  if (!match.primaryDocument || !match.filingUrl) {
    throw new CLIError(ErrorCode.NOT_FOUND, `No primary document found for accession ${accession}`);
  }

  if (params.format === 'url') {
    return {
      data: {
        accession: match.accession,
        form: match.form,
        filingDate: match.filingDate,
        reportDate: match.reportDate,
        primaryDocument: match.primaryDocument,
        url: match.filingUrl
      }
    };
  }

  const content = await context.secClient.fetchSecText(match.filingUrl);

  if (params.format === 'html') {
    return {
      data: {
        accession: match.accession,
        url: match.filingUrl,
        content
      }
    };
  }

  return {
    data: {
      accession: match.accession,
      url: match.filingUrl,
      content: extractMarkdownFromHtml(content)
    }
  };
}

File: pkg/edgar/filings.go
package edgar

import (
	"context"
	"os"
	"path/filepath"
	"strings"
)

type RecentFilings struct {
	AccessionNumber       []string `json:"accessionNumber"`
	FilingDate            []string `json:"filingDate"`
	ReportDate            []string `json:"reportDate"`
	AcceptanceDateTime    []string `json:"acceptanceDateTime"`
	Act                   []string `json:"act"`
	Form                  []string `json:"form"`
	FileNumber            []string `json:"fileNumber"`
	FilmNumber            []string `json:"filmNumber"`
	Items                 []string `json:"items"`
	Size                  []int    `json:"size"`
	IsXBRL                []int    `json:"isXBRL"`
	IsInlineXBRL          []int    `json:"isInlineXBRL"`
	PrimaryDocument       []string `json:"primaryDocument"`
	PrimaryDocDescription []string `json:"primaryDocDescription"`
}

type SubmissionsPayload struct {
	CIK     string   `json:"cik"`
	Name    string   `json:"name"`
	Tickers []string `json:"tickers"`
	Filings struct {
		Recent RecentFilings `json:"recent"`
	} `json:"filings"`
}

type FilingRow struct {
	Accession       string  `json:"accession"`
	Form            *string `json:"form"`
	FilingDate      *string `json:"filingDate"`
	ReportDate      *string `json:"reportDate"`
	PrimaryDocument *string `json:"primaryDocument"`
	FilingURL       *string `json:"filingUrl"`
}

type FilingsListParams struct {
	ID         string
	Form       string
	From       string
	To         string
	QueryLimit int
	Offset     int
}

type FilingsGetParams struct {
	ID         string
	Accession  string
	Format     string
	OutputPath string
	Raw        bool
}

func runResolve(ctx context.Context, id string, context CommandContext) (CommandResult, error) {
	entity, err := resolveEntity(ctx, id, context.SecClient, true)
	if err != nil {
		return CommandResult{}, err
	}
	return CommandResult{Data: entity}, nil
}

func zipRecentFilings(hosts SECHosts, cik string, recent RecentFilings) []FilingRow {
	rows := []FilingRow{}
	for idx, accession := range recent.AccessionNumber {
		if accession == "" {
			continue
		}
		row := FilingRow{Accession: accession}
		if idx < len(recent.Form) {
			row.Form = stringPointerOrNil(recent.Form[idx])
		}
		if idx < len(recent.FilingDate) {
			row.FilingDate = stringPointerOrNil(recent.FilingDate[idx])
		}
		if idx < len(recent.ReportDate) {
			row.ReportDate = stringPointerOrNil(recent.ReportDate[idx])
		}
		if idx < len(recent.PrimaryDocument) {
			row.PrimaryDocument = stringPointerOrNil(recent.PrimaryDocument[idx])
		}
		if row.PrimaryDocument != nil {
			if url, err := filingDocumentURL(hosts, cik, accession, *row.PrimaryDocument); err == nil {
				row.FilingURL = &url
			}
		}
		rows = append(rows, row)
	}
	return rows
}

func stringPointerOrNil(value string) *string {
	if value == "" {
		return nil
	}
	return &value
}

func runFilingsList(ctx context.Context, params FilingsListParams, context CommandContext) (CommandResult, error) {
	entity, err := resolveEntity(ctx, params.ID, context.SecClient, false)
	if err != nil {
		return CommandResult{}, err
	}
	url, err := submissionsURL(context.SecClient.Hosts(), entity.CIK)
	if err != nil {
		return CommandResult{}, err
	}

	var submissions SubmissionsPayload
	if err := context.SecClient.FetchSECJSON(ctx, url, &submissions); err != nil {
		return CommandResult{}, err
	}
	rows := zipRecentFilings(context.SecClient.Hosts(), entity.CIK, submissions.Filings.Recent)

	normalizedForm := strings.ToUpper(strings.TrimSpace(params.Form))
	filtered := []FilingRow{}
	for _, row := range rows {
		if normalizedForm != "" {
			rowForm := ""
			if row.Form != nil {
				rowForm = strings.ToUpper(*row.Form)
			}
			if rowForm != normalizedForm {
				continue
			}
		}
		if row.FilingDate == nil {
			if params.From != "" || params.To != "" {
				continue
			}
		} else if !dateInRange(*row.FilingDate, params.From, params.To) {
			continue
		}
		filtered = append(filtered, row)
	}

	offset := params.Offset
	if offset > len(filtered) {
		offset = len(filtered)
	}
	queryLimit := params.QueryLimit
	if queryLimit == 0 {
		queryLimit = len(filtered)
	}
	end := offset + queryLimit
	if end > len(filtered) {
		end = len(filtered)
	}
	pagedRows := filtered[offset:end]

	return CommandResult{
		Data: pagedRows,
		MetaUpdates: map[string]any{
			"query_total_count":    len(filtered),
			"query_returned_count": len(pagedRows),
			"query_truncated":      offset+len(pagedRows) < len(filtered),
			"query_offset":         params.Offset,
		},
	}, nil
}

func runFilingsGet(ctx context.Context, params FilingsGetParams, context CommandContext) (CommandResult, error) {
	accession, err := normalizeAccession(params.Accession)
	if err != nil {
		return CommandResult{}, err
	}
	entity, err := resolveEntity(ctx, params.ID, context.SecClient, false)
	if err != nil {
		return CommandResult{}, err
	}
	url, err := submissionsURL(context.SecClient.Hosts(), entity.CIK)
	if err != nil {
		return CommandResult{}, err
	}

	var submissions SubmissionsPayload
	if err := context.SecClient.FetchSECJSON(ctx, url, &submissions); err != nil {
		return CommandResult{}, err
	}
	rows := zipRecentFilings(context.SecClient.Hosts(), entity.CIK, submissions.Filings.Recent)
	var match *FilingRow
	for idx := range rows {
		if rows[idx].Accession == accession {
			match = &rows[idx]
			break
		}
	}
	if match == nil {
		return CommandResult{}, NewCLIError(ErrorNotFound, "Accession "+accession+" not found in recent submissions for "+params.ID)
	}
	if match.PrimaryDocument == nil || match.FilingURL == nil {
		return CommandResult{}, NewCLIError(ErrorNotFound, "No primary document found for accession "+accession)
	}

	data := map[string]any{
		"accession":       match.Accession,
		"form":            match.Form,
		"filingDate":      match.FilingDate,
		"reportDate":      match.ReportDate,
		"primaryDocument": match.PrimaryDocument,
		"url":             match.FilingURL,
	}
	if params.Format == "url" {
		return finishFilingsGet(params, data, *match.FilingURL, false)
	}

	content, err := context.SecClient.FetchSECText(ctx, *match.FilingURL)
	if err != nil {
		return CommandResult{}, err
	}
	if params.Format == "html" {
		return finishFilingsGet(params, data, content, true)
	}
	if params.Format == "text" {
		return finishFilingsGet(params, data, extractPlainTextFromHTML(content), true)
	}
	return finishFilingsGet(params, data, extractMarkdownFromHTML(content), true)
}

func finishFilingsGet(params FilingsGetParams, data map[string]any, content string, includeContent bool) (CommandResult, error) {
	if strings.TrimSpace(params.OutputPath) != "" {
		outputPath := absPath(params.OutputPath)
		outputDir := filepath.Dir(outputPath)
		if info, err := os.Stat(outputDir); err != nil {
			return CommandResult{}, NewCLIError(ErrorValidationRequired, "Unable to write output file "+outputPath+": directory does not exist")
		} else if !info.IsDir() {
			return CommandResult{}, NewCLIError(ErrorValidationRequired, "Unable to write output file "+outputPath+": parent is not a directory")
		}
		contentBytes := []byte(content)
		if err := os.WriteFile(outputPath, contentBytes, 0o644); err != nil {
			return CommandResult{}, NewCLIError(ErrorValidationRequired, "Unable to write output file "+outputPath+": "+err.Error())
		}
		data["outputPath"] = outputPath
		data["bytesWritten"] = len(contentBytes)
		data["contentFormat"] = params.Format
		return CommandResult{Data: data}, nil
	}

	if params.Raw {
		return CommandResult{RawOutput: &RawOutput{Content: content}}, nil
	}

	if includeContent {
		data["content"] = content
	}
	return CommandResult{Data: data}, nil
}
